File: typescript/session2/td/agregation/NatCalculantAvecDesNombresDecimaux.ts
import { NatDeleguantEtat } from "./NatDeleguantEtat";
import { EtatNaturelPur } from "./EtatNaturelPur";
import { Nat } from "./Nat";

export class NatCalculantAvecDesNombresDecimaux extends NatDeleguantEtat {
    constructor(_etat : EtatNaturelPur) {
        super(_etat);
    }

    creerNatAvecEtat(etat : EtatNaturelPur) : Nat {
        return new NatCalculantAvecDesNombresDecimaux(etat);
    }

    somme(x: Nat): Nat {
        const t = this.taille() < x.taille() ? x.taille() : this.taille();
        const rep: string[] = [];
        let retenue = 0;
        for (let i = 0; i < t; i++) {
            let chiffre = this.chiffre(i) + x.chiffre(i) + retenue;
            if (chiffre > 9) {
                chiffre = chiffre - 10;
                retenue = 1;
            } else {
                retenue = 0;
            }
            rep.push(chiffre.toString());
        }
        if (retenue === 1) {
            rep.push("1");
        }
        return this.creerNatAvecRepresentation(rep.reverse().join(""));
    }

    zero(): Nat {
        return this.creerZero();
    }

    produit(x: Nat): Nat {
        if (x.estEgal(this.creerNatAvecRepresentation("10"))) {
            return this.creerNatAvecRepresentation(this.representation() + "0");
        }
        let res = this.zero();
        let index = this.zero();
        while (!index.estEgal(x)) {
            res = res.somme(this);
            index = this.creerSuccesseur(index);
        }
        return res;
    }

    un(): Nat {
        return this.creerNatAvecRepresentation("1");
    }

    modulo(x: Nat): Nat {
        if (x.estEgal(this.creerNatAvecRepresentation("10"))) {
            return this.creerNatAvecValeur(this.chiffre(0));
        }
        let courant = this.zero();
        let r = this.zero();
        while (!courant.estEgal(this)) {
            r = this.creerSuccesseur(r);
            if (r.estEgal(x)) {
                r = this.zero();
            }
            courant = this.creerSuccesseur(courant);
        }
        return r;
    }

    div(x: Nat): Nat {
        if (x.estEgal(this.creerNatAvecRepresentation("10"))) {
            if (this.taille() == 1)
                return this.zero();
            return this.creerNatAvecRepresentation(
                this.representation().substring(0, this.taille() - 1));
        }
        let courant = this.zero();
        let q = this.zero();
        let r = this.zero();
        while (!courant.estEgal(this)) {
            r = this.creerSuccesseur(r);
            if (r.estEgal(x)) {
                r = this.zero();
                q = this.creerSuccesseur(q);
            }
            courant = this.creerSuccesseur(courant);
        }
        return q;
    }

    representation(): string {
        return String(this.val());
    }

    estEgal(n: Nat): boolean {
        return this.representation() === n.representation();
    }
}

File: java/cours/src/session2/td/agregation/NatCalculantAvecDesNombresDecimaux.java
package session2.td.agregation;

public class NatCalculantAvecDesNombresDecimaux extends NatDeleguantEtat implements Nat {
	
	
	public NatCalculantAvecDesNombresDecimaux(EtatNaturelPur etat) {
		super(etat);
	}

	@Override
	public Nat creerNatAvecEtat(EtatNaturelPur etat) {
		return new NatCalculantAvecDesNombresDecimaux(etat);
	}

	@Override
	public Nat somme(Nat x) {
		int t = this.taille() < x.taille() ? x.taille() : this.taille();
		StringBuilder rep = new StringBuilder();
		int retenue = 0;
		for (int i = 0; i < t; i++) {
			int chiffre = this.chiffre(i) + x.chiffre(i) + retenue;
			if (chiffre > 9) {
				chiffre = chiffre - 10;
				retenue = 1;
			} else {
				retenue = 0;
			}
			rep.append(Integer.toString(chiffre));
		}
		rep = retenue == 0 ? rep : rep.append(1);
		return this.creerNatAvecRepresentation(rep.reverse().toString());
	}

	@Override
	public Nat zero() {
		return this.creerZero();
	}

	@Override
	public Nat produit(Nat x) {
		Nat dix = this.creerNatAvecRepresentation("10");
		if (x.equals(dix)) {
			return this.creerNatAvecRepresentation(this.toString() + "0");
		}
		Nat res = zero();
		Nat index = zero();
		while (!index.equals(x)) {
			res = res.somme(this);
			index = this.creerSuccesseur(index);
		}
		return res;
	}

	@Override
	public Nat un() {
		return this.creerNatAvecRepresentation("1");
	}

	@Override
	public Nat modulo(Nat x) {
		Nat dix = this.creerNatAvecRepresentation("10");
		if (x.equals(dix)) {
			return this.creerNatAvecValeur(this.chiffre(0));
		}
		Nat courant = zero();
		Nat q = zero();
		Nat r = zero();
		while (!courant.equals(this)) {
			r = this.creerSuccesseur(r);
			if (r.equals(x)) {
				r = zero();
				q = this.creerSuccesseur(q);
			}
			courant = this.creerSuccesseur(courant);
		}
		return r;
	}

	@Override
	public Nat div(Nat x) {
		Nat dix = this.creerNatAvecRepresentation("10");
		if (x.equals(dix)) {
			if (this.taille() == 1)
				return this.zero();
			return this.creerNatAvecRepresentation(this.toString().substring(0, this.taille() - 1));
		}
		Nat courant = zero();
		Nat q = zero();
		Nat r = zero();
		while (!courant.equals(this)) {
			r = this.creerSuccesseur(r);
			if (r.equals(x)) {
				r = zero();
				q = this.creerSuccesseur(q);
			}
			courant = this.creerSuccesseur(courant);
		}
		return q;
	}


	public boolean equals(Object x) {
		if (!(x instanceof Nat))
			return false;
		Nat n = (Nat) x;
		return this.toString().equals(n.toString());
	}

	public String toString() {
		StringBuilder rep = new StringBuilder();
		for(int i = this.taille(); i > 0; i--){
			rep.append(this.chiffre(i-1));
		}
		return rep.toString();
	}

}
